File: src/shaders.cpp
#include <fstream>
#include <iostream>
#include <ranges>
#include <sstream>
#include "errors.hpp"

#include "shaders.hpp"

using shaders::ShaderProgramSource;

using std::string;
using std::vector;

namespace ranges = std::ranges;

uint32_t shaders::compileShader(uint32_t type, const vector<string>& sources) {
    uint32_t id = glCreateShader(type);

    auto raw_sources = sources
        | ranges::views::transform([](const string& s) { return s.c_str(); })
        | ranges::to<vector<const char*>>();

    CALL_GL(glShaderSource(id, raw_sources.size(), raw_sources.data(), nullptr));
    CALL_GL(glCompileShader(id));

    int32_t result;
    CALL_GL(glGetShaderiv(id, GL_COMPILE_STATUS, &result));

    if (result == GL_FALSE) {
        int32_t length;
        CALL_GL(glGetShaderiv(id, GL_INFO_LOG_LENGTH, &length));
        auto message = static_cast<char*>(alloca(sizeof(char) * static_cast<size_t>(length)));
        CALL_GL(glGetShaderInfoLog(id, length, &length, message));
        println(std::cerr);
        println(std::cerr, "Shader Compilation Error on shader type: {}", type == GL_VERTEX_SHADER ? "VERTEX" : "FRAGMENT");
        println(std::cerr, "Error: {}", message);
        println(std::cerr);
    }

    return id;
}

uint32_t shaders::compileShader(uint32_t type, const string& source) {
    const vector<string> sources = { source };
    return compileShader(type, sources);
}

uint32_t shaders::createShader(const ShaderProgramSource source) {
    uint32_t program = glCreateProgram();

    uint32_t vs = compileShader(GL_VERTEX_SHADER, source.vertex_source);
    uint32_t fs = compileShader(GL_FRAGMENT_SHADER, source.fragment_source);
    CALL_GL(glAttachShader(program, vs));
    CALL_GL(glAttachShader(program, fs));
    CALL_GL(glLinkProgram(program));
    CALL_GL(glValidateProgram(program));

    CALL_GL(glDeleteShader(vs));
    CALL_GL(glDeleteShader(fs));

    return program;
}

ShaderProgramSource shaders::parseShader(const string& file_path) {
    std::ifstream stream(file_path);

    enum class ShaderType {
        none = -1,
        vertex = 0,
        fragment = 1,
    };

    string line;

    std::stringstream shaders[] = { std::stringstream(), std::stringstream() };

    ShaderType type = ShaderType::none;
    while (getline(stream, line)) {
        if (line.find("#shader") != string::npos) {
            if (line.find("vertex") != string::npos) {
                type = ShaderType::vertex;
            } else if (line.find("fragment") != string::npos) {
                type = ShaderType::fragment;
            }
            continue;
        }

        if (type == ShaderType::none) {
            continue;
        }

        shaders[static_cast<size_t>(type)] << line << "\n";
    }

    return {
        .vertex_source = shaders[0].str(),
        .fragment_source = shaders[1].str(),
    };
}

File: src/zig/shaders.zig
const std = @import("std");
const ArrayList = std.ArrayList;
const Allocator = std.mem.Allocator;
const File = std.fs.File;

const gl = @import("gl.zig").gl;

const ShaderProgramSource = struct {
    vertex_source: [:0]const u8,
    fragment_source: [:0]const u8,

    pub fn deinit(self: ShaderProgramSource, alloc: Allocator) void{
        alloc.free(self.fragment_source);
        alloc.free(self.vertex_source);
    }
};

pub fn compileShaders(alloc: Allocator, shaderType: u32, sources: [][]const u8) Allocator.Error!u32 {
    const id = gl.glCreateShader().?(shaderType);

    const c_sources = try alloc.alloc([*c]const u8, sources.len);
    defer alloc.free(c_sources);

    for (0..c_sources.len) |i| c_sources[i] = sources[i].ptr;

    gl.glShaderSource().?(id, @intCast(c_sources.len), c_sources.ptr, null);
    gl.glCompileShader().?(id);

    var result: i32 = undefined;
    gl.glGetShaderiv().?(id, gl.GL_COMPILE_STATUS, &result);

    if (result == gl.GL_FALSE) {
        var length: i32 = undefined;
        gl.glGetShaderiv().?(id, gl.GL_INFO_LOG_LENGTH, &length);

        const message = try alloc.alloc(u8, @intCast(length));
        defer alloc.free(message);

        gl.glGetShaderInfoLog().?(id, length, &length, message.ptr);
        std.log.err(
            "\nShader Compilation Error on shader type: {s}\n",
            .{
                if (shaderType == gl.GL_VERTEX_SHADER) "VERTEX" else "FRAGMENT",
            },
        );
        std.log.err("Error: {s}\n\n", .{message});
    }

    return id;
}

pub fn compileShader(alloc: Allocator, shaderType: u32, source: []const u8) Allocator.Error!u32 {
    var sources = [_][]const u8{source};
    return compileShaders(alloc, shaderType, &sources);
}

pub fn createShader(alloc: Allocator, source: ShaderProgramSource) Allocator.Error!u32 {
    const program: u32 = gl.glCreateProgram().?();

    const vs: u32 = try compileShader(alloc, gl.GL_VERTEX_SHADER, source.vertex_source);
    const fs: u32 = try compileShader(alloc, gl.GL_FRAGMENT_SHADER, source.fragment_source);
    gl.glAttachShader().?(program, vs);
    gl.glAttachShader().?(program, fs);
    gl.glLinkProgram().?(program);
    gl.glValidateProgram().?(program);

    gl.glDeleteShader().?(vs);
    gl.glDeleteShader().?(fs);

    return program;
}

pub const ParserError = (File.OpenError || Allocator.Error || std.Io.Reader.DelimiterError);

/// Returns struct containing two owned string reads of file at file_path
/// struct members can be freed by calling: 
///
///     ShaderProgramSource.deinit(alloc);
///
pub fn parseShader(alloc: Allocator, file_path: []const u8) ParserError!ShaderProgramSource {
    const file = try std.fs.cwd().openFile(file_path, .{ .mode = .read_only });
    defer file.close();

    var read_buffer: [1024]u8 = undefined;
    var file_reader = file.reader(&read_buffer);

    const ShaderType = enum {
        vertex,
        fragment,
    };

    var shaders: std.EnumArray(ShaderType, ArrayList(u8)) = .init(.{
        .vertex = .empty,
        .fragment = .empty,
    });

    var shaderType: ?ShaderType = null;

    while (try file_reader.interface.takeDelimiter('\n')) |line| {
        if (std.mem.indexOf(u8, line, "#shader")) |_| {
            if (std.mem.indexOf(u8, line, "vertex")) |_| {
                shaderType = .vertex;
            } else if (std.mem.indexOf(u8, line, "fragment")) |_| {
                shaderType = .fragment;
            }
            continue;
        }

        if (shaderType) |t| {
            try shaders.getPtr(t).appendSlice(alloc, line);
            try shaders.getPtr(t).append(alloc, '\n');
        }
    }

    return .{
        .vertex_source = try shaders.getPtr(.vertex).toOwnedSliceSentinel(alloc, 0),
        .fragment_source = try shaders.getPtr(.fragment).toOwnedSliceSentinel(alloc, 0),
    };
}
